lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -0,0 +1,151 @@
+        # string. These aren't automatically validated by the superclass, so we
+        # have to do it here.
+        try:
+            self.logger.info("pre-validate {!r}", json)
+            new_json = self.schema.validate(json)
+            self.logger.info("post-validate {!r}", new_json)
+        except SchemaError as e:
+            self.logger.warning("Schema validation error {}", e)
+            abort(HTTPStatus.BAD_REQUEST, message=str(e))
+
+        # Validate the authenticated user's authorization for the combination
+        # of "owner" and "access".
+        self._check_authorization(json_data.get("owner"), new_json.get("access"))
+
+        # Build a SQLAlchemy Query object expressing all of our constraints
+        #
+        # TODO: Think about what other constraints we might want to support.
+        # Including Metadata outside the Dataset table would be more difficult,
+        # but should be possible.
+        query = Database.db_session.query(Dataset)
+        if "start" in new_json and "end" in new_json:
+            self.logger.info("Adding start / end query")
+            query = query.filter(
+                Dataset.created.between(new_json["start"], new_json["end"])
+            )
+        elif "start" in new_json:
+            self.logger.info("Adding start query")
+            query = query.filter(Dataset.created >= new_json["start"])
+        elif "end" in new_json:
+            self.logger.info("Adding end query")
+            query = query.filter(Dataset.created <= new_json["end"])
+        if "name" in new_json:
+            self.logger.info("Adding name query")
+            query = query.filter(Dataset.name.contains(new_json["name"]))
+        if "controller" in new_json:
+            self.logger.info("Adding controller query")
+            query = query.filter(Dataset.controller == new_json["controller"])
+        query = self._build_sql_query(new_json, query)
+
+        # Useful for debugging, but verbose: this displays the fully expanded
+        # SQL `SELECT` statement.
+        self.logger.debug(
+            "QUERY {}", query.statement.compile(compile_kwargs={"literal_binds": True})
+        )
+
+        # Execute the filtered query, sorted by dataset name so we have a
+        # consistent and reproducible output to compare.
+        results = query.order_by(Dataset.name).all()
+
+        keys = new_json.get("metadata")
+
+        response = []
+        for dataset in results:
+            # TODO: Decide what we really need here for basic information. We
+            # should probably drop "controller" but allow "dataset.controller"
+            # metadata. We currently require "run_id" for the TOC, iteration
+            # sample, and timeseries APIs, although those *should* be based on
+            # dataset name. Should the output here ultimately be anything but
+            # a list of matching dataset names and requested metadata for each?
```

Comment:
Yeah, you have full control of the metadata returned, and I'd like to have the entire `Dataset` schema reflected in the `dataset.` namespace (though it's not yet), so, arguably, just a list of names would be fine. (But, again, I haven't exposed the run_it/md5 yet, which _is_ necessary for the other APIs at this point... though adding `dataset.md5` would allow addressing that later without changing the API definition, and I do kinda like that so maybe that's worth doing.)

At the same time, while I haven't wanted to derail Riya's work on TOC, changing all of those "second level" info APIs to be based on dataset name instead of run_id would make a lot more sense going forward.